### PR TITLE
feat(lance): Add distributed IVF vector index creation with Daft

### PR DESCRIPTION
**Pull request by huleilei**
## Changes Made
This PR implements distributed IVF (Inverted File) vector index creation functionality for LanceDB using Daft's distributed computing capabilities. The implementation enables efficient building of vector indices on large-scale datasets by leveraging Daft's parallel processing framework.

1. New create_index Public API
   
   - Provides a unified entry point for vector index creation
   - Supports multiple IVF vector index types: IVF_PQ , IVF_HNSW_PQ , IVF_FLAT , IVF_SQ , IVF_HNSW_SQ , IVF_HNSW_FLAT
   - Supports multiple distance metrics: l2 , cosine , euclidean , dot , hamming
2. Distributed Index Building
   
   - Implements distributed index building using Daft's UDF engine
   - Supports custom concurrency parameters for performance optimization
   - Automatically computes fragment_group_size and repartitions based on concurrency
   - Supports GPU acceleration via daft_remote_args configuration
3. Intelligent Parameter Configuration
   
   - Automatically determines IVF partition count based on dataset size
   - Automatically determines PQ sub-vector count based on vector dimension
   - Supports pre-trained PQ codebooks
4. Comprehensive Test Coverage
   
   - Adds 387 lines of test code covering 12 test cases
   - Tests different index types, distance metrics, and concurrency settings
   - Validates edge cases such as index replacement and custom naming

## Related Issues
https://github.com/Eventual-Inc/Daft/discussions/6375

### CODE REVIEW

File: daft/io/lance/lance_vector_index.py [hunk 1]
```diff
@@ -0,0 +1,381 @@
...
+    index_id = str(uuid.uuid4())
+    logger.info(
+        "Starting distributed vector index creation: column=%s, type=%s, name=%s, metric=%s, ivf_num_partitions=%s",
+        column,
+        index_type_name,
+        name,
+        metric_lower,
+        concurrency,
+        ivf_num_partitions,
+    )
```
[the remark this answers is not in the document]
Comment (huleilei) [reply daft/io/lance/lance_vector_index.py:273]: Good catch. The log format string was missing the concurrency placeholder, so ivf_num_partitions was not logged correctly. Fixed by logging both concurrency and ivf_num_partitions.

File: daft/io/lance/_lance.py [hunk 2]
```diff
@@ -514,6 +514,112 @@ def create_scalar_index(
...
+        If the ``lance`` package is not available.
+    """
+    try:
+        import lance  # type: ignore[import-not-found]
+
+        from daft.io.lance.lance_vector_index import (
+            _check_pylance_version,
+            _normalize_index_type,
+            create_vector_index_internal,
+        )
```
[the remark this answers is not in the document]
Comment (huleilei) [reply daft/io/lance/_lance.py]: I’m keeping the lazy import pattern here for now because lance is still treated as an optional dependency in this module, and importing at function scope avoids import-time failures for users without the lance extra installed. Happy to revisit if we want to standardize optional-dependency handling across _lance.py.

File: daft/io/lance/lance_vector_index.py [hunk 3]
```diff
@@ -0,0 +1,381 @@
...
+    create_index_op = lance.LanceOperation.CreateIndex(
+        new_indices=[index],
+        removed_indices=[],
+    )
+    LanceDataset.commit(
+        uri,
+        create_index_op,
+        read_version=lance_ds.version,
+        storage_options=storage_options,
+    )
```
[the remark this answers is not in the document]
Comment (huleilei) [reply daft/io/lance/lance_vector_index.py:386]: I don’t think this is necessarily a bug. In the distributed path, the worker-side call already delegates `replace=self.replace` into `LanceDataset.create_index(...)`, so the replace semantics are primarily handled by Lance itself during index creation and metadata merge. The final `LanceOperation.CreateIndex(new_indices=[...], removed_indices=[...]) `appears to be a lower-level commit representation, and without confirming Lance’s internal implementation we shouldn’t assume that `removed_indices` must be populated manually for` replace=True` to work. Given that, I’d prefer not to change this part unless we can validate the exact Lance-side contract.

**Comment by huleilei** [comment]
@plotor When you have time, please help review the code, thank you.

**Comment by rchowell** [comment]
@huleilei thanks for the contribution. Could you please open this PR against https://github.com/daft-engine/daft-lance?

We are moving our Lance-specific features here, thank you!
